src/librustc_mir/interpret/eval_context.rs

```diff
@@ -780,7 +793,11 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
                 let ty = self.place_ty(place);
                 let place = self.eval_place(place)?;
                 let discr_val = self.read_discriminant_value(place, ty)?;
-                self.write_primval(dest, PrimVal::Bytes(discr_val), dest_ty)?;
+                let defined = self.layout_of(ty).unwrap().size.bits() as u8;
```

Comment:
I believe this causes #51086. `ty` is the enum of which the discriminant is taken. This should have been `dest_ty`